File: app/src/main/java/com/trevorwiebe/trackacow/presentation/activities/AddDrugsGivenToSpecificCowActivity.java
package com.trevorwiebe.trackacow.presentation.activities;

import com.google.android.material.snackbar.Snackbar;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;

import androidx.cardview.widget.CardView;
import android.text.InputType;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.data.cacheEntities.CacheDrugsGivenEntity;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.drugsGiven.InsertDrugsGivenList;
import com.trevorwiebe.trackacow.domain.dataLoaders.cache.holdingDrugsGiven.InsertHoldingDrugsGivenList;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.drug.QueryAllDrugs;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.cow.QueryCowIdByCowId;
import com.trevorwiebe.trackacow.data.entities.CowEntity;
import com.trevorwiebe.trackacow.data.entities.DrugEntity;
import com.trevorwiebe.trackacow.data.entities.DrugsGivenEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.Utility;

import java.util.ArrayList;

public class AddDrugsGivenToSpecificCowActivity extends AppCompatActivity implements
        QueryAllDrugs.OnAllDrugsLoaded,
        QueryCowIdByCowId.OnCowByIdLoaded {


    private ArrayList<DrugEntity> mDrugList = new ArrayList<>();
    private DatabaseReference mBaseRef;
    private String mCowId;
    private CowEntity mCowEntity;

    private LinearLayout mDrugLayout;
    private TextView mNoDrugs;
    private Button mCancelButton;
    private Button mSaveButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_drugs_given_to_specific_cow);

        mBaseRef = FirebaseDatabase.getInstance().getReference("users").child(FirebaseAuth.getInstance().getCurrentUser().getUid());

        this.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);
        this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mCowId = getIntent().getStringExtra("cowId");

        new QueryAllDrugs(this).execute(this);
        new QueryCowIdByCowId(AddDrugsGivenToSpecificCowActivity.this, mCowId).execute(AddDrugsGivenToSpecificCowActivity.this);

        mDrugLayout = findViewById(R.id.drug_layout);
        mNoDrugs = findViewById(R.id.no_drugs_added_to_specific_cow);
        mCancelButton = findViewById(R.id.add_drug_cancel_button);
        mSaveButton = findViewById(R.id.add_drug_save_button);

        mCancelButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setResult(RESULT_CANCELED);
                finish();
            }
        });
        mSaveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                DatabaseReference drugsGivenRef = mBaseRef.child(Constants.DRUGS_GIVEN);

                ArrayList<DrugsGivenEntity> drugList = new ArrayList<>();

                for (int r = 0; r < mDrugLayout.getChildCount(); r++) {
                    DrugsGivenEntity drugsGivenEntity = new DrugsGivenEntity();
                    drugsGivenEntity.setDrugsGivenCowId(mCowEntity.getCowId());

                    View cardView = mDrugLayout.getChildAt(r);

                    if (cardView instanceof CardView) {

                        View linearLayout = ((CardView) cardView).getChildAt(0);

                        if (linearLayout instanceof LinearLayout) {

                            LinearLayout confirmedLinearLayout = (LinearLayout) linearLayout;

                            View checkBoxView = confirmedLinearLayout.getChildAt(0);
                            if (checkBoxView instanceof CheckBox) {

                                CheckBox checkBox = (CheckBox) checkBoxView;
                                String drugId = checkBox.getTag().toString().split("#")[0];
                                drugsGivenEntity.setDrugsGivenDrugId(drugId);

                                if (checkBox.isChecked()) {

                                    View editText = confirmedLinearLayout.getChildAt(2);

                                    if (editText instanceof EditText) {

                                        EditText textViewAmountGiven = (EditText) editText;
                                        int amountGiven = Integer.parseInt(textViewAmountGiven.getText().toString());
                                        drugsGivenEntity.setDrugsGivenAmountGiven(amountGiven);

                                        DatabaseReference drugsGivenPushRef = drugsGivenRef.push();
                                        String drugsGivenKey = drugsGivenPushRef.getKey();
                                        drugsGivenEntity.setDrugsGivenLotId(mCowEntity.getLotId());
                                        drugsGivenEntity.setDrugsGivenId(drugsGivenKey);
                                        drugsGivenEntity.setDrugsGivenDate(mCowEntity.getDate());

                                        drugList.add(drugsGivenEntity);

                                    }
                                }
                            }
                        }
                    }
                }

                if (drugList.size() == 0) {
                    Snackbar.make(v, "You must select a drug before saving", Snackbar.LENGTH_LONG).show();
                    return;
                }

                if (Utility.haveNetworkConnection(AddDrugsGivenToSpecificCowActivity.this)) {

                    for (int k = 0; k < drugList.size(); k++) {
                        DrugsGivenEntity drugsGivenEntity = drugList.get(k);
                        drugsGivenRef.child(drugsGivenEntity.getDrugsGivenId()).setValue(drugsGivenEntity);
                    }

                } else {

                    Utility.setNewDataToUpload(AddDrugsGivenToSpecificCowActivity.this, true);

                    // array list to hold the holdingDrugsGivenEntities so we can push them all at once to the local db
                    ArrayList<CacheDrugsGivenEntity> holdingDrugsGivenEntities = new ArrayList<>();

                    // iterate over the drugGivenEntityList
                    for (int q = 0; q < drugList.size(); q++) {
                        DrugsGivenEntity drugsGivenEntity = drugList.get(q);

                        CacheDrugsGivenEntity cacheDrugsGivenEntity = new CacheDrugsGivenEntity();
                        cacheDrugsGivenEntity.setAmountGiven(drugsGivenEntity.getDrugsGivenAmountGiven());
                        cacheDrugsGivenEntity.setCowId(drugsGivenEntity.getDrugsGivenCowId());
                        cacheDrugsGivenEntity.setDrugId(drugsGivenEntity.getDrugsGivenDrugId());
                        cacheDrugsGivenEntity.setDrugGivenId(drugsGivenEntity.getDrugsGivenId());
                        cacheDrugsGivenEntity.setWhatHappened(Constants.INSERT_UPDATE);
                        cacheDrugsGivenEntity.setLotId(drugsGivenEntity.getDrugsGivenLotId());

                        holdingDrugsGivenEntities.add(cacheDrugsGivenEntity);
                    }

                    new InsertHoldingDrugsGivenList(holdingDrugsGivenEntities).execute(AddDrugsGivenToSpecificCowActivity.this);

                }

                new InsertDrugsGivenList(drugList).execute(AddDrugsGivenToSpecificCowActivity.this);

                setResult(RESULT_OK);
                finish();

            }
        });

    }

    @Override
    public void onAllDrugsLoaded(ArrayList<DrugEntity> drugEntities) {
        mDrugList = drugEntities;
        for (int x = 0; x < mDrugList.size(); x++) {
            DrugEntity drugEntity = drugEntities.get(x);
            addCheckBox(mDrugLayout, drugEntity);
        }
        if (drugEntities.size() == 0) {
            mNoDrugs.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onCowByIdLoaded(CowEntity cowEntity) {
        mCowEntity = cowEntity;
        String tagNumber = Integer.toString(mCowEntity.getTagNumber());
        setTitle("Medicate cow " + tagNumber);
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return false;
    }

    private void addCheckBox(LinearLayout linearLayout, DrugEntity drugEntity) {

        String drugName = drugEntity.getDrugName();
        String drugId = drugEntity.getDrugCloudDatabaseId();
        int defaultAmount = drugEntity.getDefaultAmount();
        String defaultAmountStr = Integer.toString(defaultAmount);

        final float scale = getResources().getDisplayMetrics().density;
        int pixels24 = (int) (24 * scale + 0.5f);
        int pixels16 = (int) (16 * scale + 0.5f);
        int pixels8 = (int) (8 * scale + 0.5f);

        CardView cardView = new CardView(this);
        LinearLayout.LayoutParams cardViewParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        cardViewParams.setMargins(pixels8, pixels8, pixels8, pixels8);
        cardView.setLayoutParams(cardViewParams);
        cardView.setTag(drugId + "#cardView");
        cardView.setOnClickListener(cardViewClickListener);

        LinearLayout containerLayout = new LinearLayout(this);
        LinearLayout.LayoutParams containerParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        containerLayout.setTag(drugId + "#linearLayout");
        containerLayout.setOrientation(LinearLayout.HORIZONTAL);
        containerLayout.setLayoutParams(containerParams);

        cardView.addView(containerLayout);

        CheckBox checkBox = new CheckBox(this);
        checkBox.setText(drugName);
        checkBox.setTag(drugId);
        LinearLayout.LayoutParams checkBoxParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        checkBoxParams.setMargins(pixels24, pixels8, pixels24, pixels8);
        checkBox.setTag(drugId + "#checkBox");
        checkBox.setOnCheckedChangeListener(checkedChangeListener);
        checkBox.setLayoutParams(checkBoxParams);

        View view = new View(this);
        LinearLayout.LayoutParams viewParams = new LinearLayout.LayoutParams(
                0,
                0,
                1
        );
        view.setLayoutParams(viewParams);

        EditText editText = new EditText(this);
        LinearLayout.LayoutParams editTextParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        editTextParams.setMargins(pixels24, pixels8, pixels24, pixels8);
        editText.setEms(4);
        editText.setGravity(Gravity.CENTER);
        editText.setTag(drugId + "#editText");
        editText.setText(defaultAmountStr);
        editText.setSelectAllOnFocus(true);
        editText.setInputType(InputType.TYPE_CLASS_NUMBER);
        editText.setLayoutParams(editTextParams);

        containerLayout.addView(checkBox);
        containerLayout.addView(view);
        containerLayout.addView(editText);

        linearLayout.addView(cardView);
    }

    View.OnClickListener cardViewClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            String cardViewTag = v.getTag().toString();
            String drugId = cardViewTag.split("#")[0];

            LinearLayout linearLayout = v.findViewWithTag(drugId + "#linearLayout");
            CheckBox checkBox = linearLayout.findViewWithTag(drugId + "#checkBox");
            if (checkBox.isChecked()) {
                checkBox.setChecked(false);
                v.setBackgroundColor(getResources().getColor(android.R.color.white));
            } else {
                checkBox.setChecked(true);
                v.setBackgroundColor(getResources().getColor(R.color.colorAccentVeryLight));
            }
        }
    };

    CheckBox.OnCheckedChangeListener checkedChangeListener = new CompoundButton.OnCheckedChangeListener() {
        @Override
        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
            LinearLayout linearLayout = (LinearLayout) buttonView.getParent();
            CardView cardView = (CardView) linearLayout.getParent();
            if (isChecked) {
                cardView.setBackgroundColor(getResources().getColor(R.color.colorAccentVeryLight));
            } else {
                cardView.setBackgroundColor(getResources().getColor(android.R.color.white));
            }
        }
    };
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/add_drugs_given_to_specific_cow/AddDrugsGivenToSpecificCowActivity.kt
package com.trevorwiebe.trackacow.presentation.add_drugs_given_to_specific_cow

import android.os.Build
import android.os.Bundle
import android.text.InputType
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.google.android.material.snackbar.Snackbar
import com.trevorwiebe.trackacow.R
import com.trevorwiebe.trackacow.domain.models.cow.CowModel
import com.trevorwiebe.trackacow.domain.models.drug.DrugModel
import com.trevorwiebe.trackacow.domain.models.drug_given.DrugGivenModel
import com.trevorwiebe.trackacow.domain.utils.DataSource
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class AddDrugsGivenToSpecificCowActivity : AppCompatActivity() {

    private val mAddDrugsGivenToSpecificCowViewModel: AddDrugsGivenToSpecificCowViewModel by viewModels()

    private var mDrugList: MutableList<DrugModel> = mutableListOf()
    private var mCowModel: CowModel? = null

    private lateinit var mDrugLayout: LinearLayout
    private lateinit var mNoDrugs: TextView
    private lateinit var mCancelButton: Button
    private lateinit var mSaveButton: Button
    private lateinit var mProgressIndicator: LinearProgressIndicator

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_drugs_given_to_specific_cow)

        this.supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        this.supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        @Suppress("DEPRECATION")
        mCowModel = if (Build.VERSION.SDK_INT >= 33) {
            intent.getParcelableExtra("cowModel", CowModel::class.java)
        } else {
            intent.getParcelableExtra("cowModel")
        }

        val tagNumber = mCowModel?.tagNumber.toString()
        title = "Medicate cow $tagNumber"

        mDrugLayout = findViewById(R.id.drug_layout)
        mNoDrugs = findViewById(R.id.no_drugs_added_to_specific_cow)
        mCancelButton = findViewById(R.id.add_drug_cancel_button)
        mSaveButton = findViewById(R.id.add_drug_save_button)
        mProgressIndicator = findViewById(R.id.add_drug_specific_cow_progress_indicator)
        mCancelButton.setOnClickListener {
            setResult(RESULT_CANCELED)
            finish()
        }

        mSaveButton.setOnClickListener(View.OnClickListener { v ->
            val drugList: MutableList<DrugGivenModel> = mutableListOf()
            for (r in 0 until mDrugLayout.childCount) {
                val drugGivenModel = DrugGivenModel()
                drugGivenModel.drugsGivenCowId = mCowModel?.cowId
                val cardView = mDrugLayout.getChildAt(r)
                if (cardView is CardView) {
                    val linearLayout = cardView.getChildAt(0)
                    if (linearLayout is LinearLayout) {
                        val checkBoxView = linearLayout.getChildAt(0)
                        if (checkBoxView is CheckBox) {
                            val drugId = checkBoxView.tag.toString().split("#".toRegex())
                                .dropLastWhile { it.isEmpty() }
                                .toTypedArray()[0]
                            drugGivenModel.drugsGivenDrugId = drugId
                            if (checkBoxView.isChecked) {
                                val editText = linearLayout.getChildAt(2)
                                if (editText is EditText) {
                                    val amountGiven = editText.text.toString().toInt()
                                    drugGivenModel.drugsGivenAmountGiven = amountGiven
                                    drugGivenModel.drugsGivenLotId = mCowModel?.lotId ?: ""
                                    drugGivenModel.drugsGivenDate = mCowModel?.date ?: 0
                                    drugList.add(drugGivenModel)
                                }
                            }
                        }
                    }
                }
            }

            if (drugList.size == 0) {
                Snackbar
                    .make(v, "You must select a drug before saving", Snackbar.LENGTH_LONG)
                    .show()
                return@OnClickListener
            }

            mAddDrugsGivenToSpecificCowViewModel.onEvent(
                AddDrugsGivenToSpecificCowEvent.OnDrugListCreated(
                    drugList
                )
            )

            setResult(RESULT_OK)
            finish()
        })


        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                mAddDrugsGivenToSpecificCowViewModel.uiState.collect {

                    if (it.isFetchingFromCloud && it.dataSource == DataSource.Local) {
                        mProgressIndicator.visibility = View.VISIBLE
                    } else {
                        mProgressIndicator.visibility = View.GONE
                    }

                    mDrugLayout.removeAllViews()

                    mDrugList = it.drugsList.toMutableList()
                    for (x in mDrugList.indices) {
                        val drugEntity = mDrugList[x]
                        addCheckBox(mDrugLayout, drugEntity)
                    }
                    if (mDrugList.size == 0) {
                        mNoDrugs.visibility = View.VISIBLE
                    } else {
                        mNoDrugs.visibility = View.GONE
                    }

                }
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return false
    }

    private fun addCheckBox(linearLayout: LinearLayout, drugModel: DrugModel) {
        val drugName = drugModel.drugName
        val drugId = drugModel.drugCloudDatabaseId
        val defaultAmount = drugModel.defaultAmount
        val defaultAmountStr = defaultAmount.toString()
        val scale = resources.displayMetrics.density
        val pixels24 = (24 * scale + 0.5f).toInt()
        val pixels8 = (8 * scale + 0.5f).toInt()
        val cardView = CardView(this)
        val cardViewParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        cardViewParams.setMargins(pixels8, pixels8, pixels8, pixels8)
        cardView.layoutParams = cardViewParams
        cardView.tag = "$drugId#cardView"
        cardView.radius = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            50f,
            this.resources.displayMetrics
        )
        cardView.setOnClickListener(cardViewClickListener)
        val containerLayout = LinearLayout(this)
        val containerParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        containerLayout.tag = "$drugId#linearLayout"
        containerLayout.orientation = LinearLayout.HORIZONTAL
        containerLayout.layoutParams = containerParams
        cardView.addView(containerLayout)
        val checkBox = CheckBox(this)
        checkBox.text = drugName
        checkBox.tag = drugId
        val checkBoxParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        checkBoxParams.setMargins(pixels24, pixels8, pixels24, pixels8)
        checkBox.tag = "$drugId#checkBox"
        checkBox.setOnCheckedChangeListener(checkedChangeListener)
        checkBox.layoutParams = checkBoxParams
        val view = View(this)
        val viewParams = LinearLayout.LayoutParams(
            0,
            0,
            1f
        )
        view.layoutParams = viewParams
        val editText = EditText(this)
        val editTextParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        editTextParams.setMargins(pixels24, pixels8, pixels24, pixels8)
        editText.setEms(4)
        editText.gravity = Gravity.CENTER
        editText.tag = "$drugId#editText"
        editText.setText(defaultAmountStr)
        editText.setSelectAllOnFocus(true)
        editText.inputType = InputType.TYPE_CLASS_NUMBER
        editText.layoutParams = editTextParams
        containerLayout.addView(checkBox)
        containerLayout.addView(view)
        containerLayout.addView(editText)
        linearLayout.addView(cardView)
    }

    private var cardViewClickListener = View.OnClickListener { v ->
        val cardViewTag = v.tag.toString()
        val drugId = cardViewTag.split("#".toRegex()).dropLastWhile { it.isEmpty() }
            .toTypedArray()[0]
        val linearLayout = v.findViewWithTag<LinearLayout>("$drugId#linearLayout")
        val checkBox = linearLayout.findViewWithTag<CheckBox>("$drugId#checkBox")
        checkBox.isChecked = !checkBox.isChecked
    }
    private var checkedChangeListener =
        CompoundButton.OnCheckedChangeListener { buttonView, isChecked ->
            val linearLayout = buttonView.parent as LinearLayout
            val cardView = linearLayout.parent as CardView
            if (isChecked) {
                cardView.setCardBackgroundColor(
                    ContextCompat.getColor(
                        this,
                        R.color.colorAccentVeryLight
                    )
                )
            } else {
                cardView.setCardBackgroundColor(ContextCompat.getColor(this, android.R.color.white))
            }
        }
}